subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing.junitplatform;
+
+import org.gradle.api.internal.tasks.testing.DefaultTestDescriptor;
+import org.gradle.api.internal.tasks.testing.TestDescriptorInternal;
+import org.gradle.api.internal.tasks.testing.TestResultProcessor;
+import org.gradle.api.internal.tasks.testing.junit.JUnitTestEventAdapter;
+import org.gradle.internal.id.IdGenerator;
+import org.gradle.internal.time.Clock;
+import org.junit.platform.engine.TestExecutionResult;
+import org.junit.platform.engine.TestSource;
+import org.junit.platform.engine.support.descriptor.ClassSource;
+import org.junit.platform.engine.support.descriptor.MethodSource;
+import org.junit.platform.launcher.TestExecutionListener;
+import org.junit.platform.launcher.TestIdentifier;
+import org.junit.platform.launcher.TestPlan;
+
+public class JUnitPlatformTestExecutionListener extends JUnitTestEventAdapter implements TestExecutionListener {
```

Comment:
It looks like some of the JUnit5 classes here extend the JUnit4 ones, which in turn utilize classes from the Junit4 libraries.  Does the cause problems if the project being built doesn't include them as a dependency?  In addition, would it be cleaner to separate out the shared functionality between the JUnit5 & JUnit4 classes so that they don't have cross version dependencies?

Reply:
Nice catch! In theory, if we remove the JUnit 5 dependency in test runtime classpath, the test should fail with `ClassNotFoundError`. I tried just now but it didn't fail, which made me quite puzzled. I need to investigate it and try to add a test for this scenario.

Reply:
I'll extract common logic to an `AbstractJUnitTestEventAdapter` tomorrow.

Reply:
My purpose of making JUnit 5 class extend JUnit 4 one is to reuse some code in `JUnitTestEventAdapter`. You're right, we shouldn't make this happen. However, I can't extract common code into an `AbstractJUnitTestEventAdapter` and let JUnit 4/5 extend it because JUnit 4's `JUnitTestEventAdapter` must be base class of `RunListener`. So I extract common code into [`GenericJUnitTestEventAdapter`](https://github.com/gradle/gradle/blob/39acfed6c6c8e00e3db49fc5a51f68be3b631f9c/subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/junit/GenericJUnitTestEventAdapter.java#L36) and create such field in both `JUnitTestEventAdapter` and `JUnitPlatformTestExecutionListener`.